# Python Decorators

In Python, a decorator is a special kind of function that can be used to modify the behavior of another function. Decorators are indicated by the `@` symbol followed by the name of the decorator function, and they are placed immediately before the function that they modify. Here's an example:

In [1]:
def my_decorator(func):
    def wrapper():
        print("Before function execution")
        func()
        print("After function execution.")
    return wrapper

@my_decorator
def say_hello():
    print("Hello, world!")

say_hello()

Before function execution
Hello, world!
After function execution.


In this example, we define a decorator function `my_decorator` that takes a function `func` as an argument. The decorator function defines an inner function `wrapper` that prints a message before and after calling the original function `func`. The decorator function then returns the `wrapper` function.

To apply the decorator to a function, we simply add the `@my_decorator` annotation immediately before the function definition. In this case, we are decorating the `say_hello` function. When we call `say_hello()`, the decorator function `my_decorator` is called with `say_hello` as its argument. The decorator returns the `wrapper` function, which is then used to call the original `say_hello` function.

Decorators can also take arguments. For example:

In [4]:
def repeat(n):
    def decorator(funct):
        def wrapper(*args, **kwargs):
            print(f"It will run {n} times.")
            for i in range(n):
                funct(*args, **kwargs)
        return wrapper
    return decorator

@repeat(3)
def say_hello(name):
    print(f"Hello, {name}")

say_hello('Punk')

It will run 3 times.
Hello, Punk
Hello, Punk
Hello, Punk


In this example, `repeat` is a function that takes an argument `n` and returns another function `decorator`. The `decorator` function takes a function `funct` as an argument and returns a new function `wrapper`. The `wrapper` function takes any number of positional and keyword arguments, calls `funct` with those arguments `n` times, and discards the results.

The `@repeat(3)` syntax is a shorthand way of applying the `repeat decorator` to the `say_hello` function with an argument of `3`. It's equivalent to calling `say_hello` = `repeat(3)(say_hello)`.

Notice that the `wrapper` function called `say_hello` three times.

<br>

#### Ques: Create a decorator called `timer` that calculates and prints the time taken by a function to execute. Use the `time` module to measure the execution time. Apply this decorator to a function `my_function` that simply prints "Hello, world!" and sleeps for 2 seconds using `time.sleep(2)`. Your task is to define the `timer` decorator and apply it to `my_function`, then call `my_function` to verify that the timer is working correctly.

In [2]:
import time
def timer(func):
    def wrapper():
        start = time.time()
        time.sleep(2)
        end = time.time()
        result = end - start
        print(f"Time taken: {result:.2f} seconds.")
    return wrapper

@timer
def my_function():
    print("Hello, world!")

my_function()

Time taken: 2.01 seconds.


#### Quest :Class Discussion : Check email using decorator .

In [3]:
def validation(func):
    def wrapper(email):
        is_valid = is_valid_email(email)
        if is_valid:
            print(f"{email} is a valid email address")
        else:
            print(f"{email} is not a valid email address")
    return wrapper

def is_valid_email(email):
    parts = email.split('@')
    if len(parts) != 2:
        return False
    username, domain = parts
    if not username or not domain:
        return False
    if '.' not in domain:
        return False
    return True
    

@validation
def valid_email(email):
    pass
email = input("Enter an email address: ")
valid_email(email)
    

Enter an email address:  prabinthapa982411@gmail.com


prabinthapa982411@gmail.com is a valid email address


In [23]:
def validation(func):
    def check_email(email):
        if email.endswith('@gmail.com'):
            func(email)
        else:
            print("We only support email from google.")
    return check_email

@validation
def user_login(email):
    print("Accessing DB..")
    print("Logged in successfully.")

user_login(input("Enter your email address: "))

Enter your email address:  sushmawaiba49@gmail.com


Accessing DB..
Logged in successfully.


## Python Property Decorator:

In Python, the `@property` decorator is used to define a getters, setters, and deleters method for a class attribute. They are used to ensure that the attributes of a class are accessed and modified in a controlled manner.

## Getter

A getter method is used to get the value of an attribute. It is called automatically when the attribute is accessed. In Python, we can define a getter method using the `@property` decorator.

In [5]:
class Person:
    def __init__(self, name):
        self.name = name

    @property
    def get_name(self):
        print("Getting name... ")
        return self.name

In [6]:
person = Person("Alice")
person.get_name

Getting name... 


'Alice'

In this example, `Person` is a class that has an attribute `name` and a method `name` decorated with `@property`. The `name` method simply returns the value of the `name` attribute. When we create an instance of `Person` and access the `name` attribute with `person.name`, the `@property` decorator automatically calls the `name` method and returns its result.

Notice that the `name` method was called automatically when we accessed the `name` attribute

<br>

## Setter

A setter method is used to set the value of an attribute. It is called automatically when the attribute is assigned a new value. In Python, we can define a setter method using the `@property` decorator with the `@setter` decorator.

In [10]:
class Person:
    def __init__(self, name):
        self._name = name

    @property
    def name(self):
        print("Getting name...")
        return self._name

    @name.setter
    def name(self, value):
        print("Setting name...")
        self._name = value

person = Person("Alice")
print(person.name)
        

Getting name...
Alice


In [11]:
person.name = "Bob"
print(person.name)

Setting name...
Getting name...
Bob


In this example, `Person` is a class that has an attribute `_name`, a method `name` decorated with `@property`, and a method `name` decorated with `@name.setter`. The `name` method gets and sets the `_name` attribute. When we create an instance of `Person` and access the `name` attribute with `person.name`, the `@property` decorator automatically calls the `name` method and returns its result. When we set the `name` attribute with `person.name` = `"Bob"`, the `@name.setter` decorator automatically calls the `name` method with the value of `"Bob"`.

Notice that the `name` method was called automatically when we accessed or set the `name` attribute.

<br>

## Deleter

A deleter method is used to delete an attribute. It is called automatically when the `del` statement is used to delete the attribute. In Python, we can define a `deleter` method using the `@property` decorator with the `@deleter` decorator.

In [13]:
class Person:
    def __init__(self, name):
        self._name = name

    @property
    def name(self):
        print("Getting name.....")
        return self._name

    @name.setter
    def name(self, value):
        print("Setting name")
        self._name = value

    @name.deleter
    def name(self):
        print("Deleting name....")
        del self._name

person = Person("Alice")
print(person.name)

Getting name.....
Alice


In [14]:
del person.name

Deleting name....


In [15]:
print(person.name)

Getting name.....


AttributeError: 'Person' object has no attribute '_name'

In this example, `Person` is a class that has an attribute `_name`, a method `name` decorated with `@property`, a method `name` decorated with `@name.setter`, and a method `name` decorated with `@name.deleter`. The `name` method gets, sets, and deletes the `_name attribute`. When we create an instance of `Person` and access the `name` attribute with `person.name`, the `@property` decorator automatically calls the `name` method and returns its results. When we set the `name` attribute with `person.name` = "Alice", the `@name.setter` decorator automatically calls the `name` method with the value of `"Alice"`. When we `delete` the `name` attribute with `del person.name`, the `@name.deleter` decorator automatically calls the `name` method and `deletes` the value assigned to `name` attribute. Now, when we try to access the `name` attribute with `person.name`, an `AttributeError` is raised because the attribute no longer exists.

Notice that the name method was called automatically when we accessed or set or deleted the name attribute.

<br><br><br>

# Python Iterators

An iterator is an object that can be iterated (looped) upon, meaning we can traverse through all the values in the iterator one by one. An iterator is an object that implements the `__iter__()` and `__next__()` methods.

The `__iter__()` method returns the iterator object itself. It is used to initialize the iteration, and it is called when the `iter()` function is called on an iterator object.

The `__next__()` method returns the next value in the iteration. It is called when the `next()` function is called on an iterator object.

Here is an example of an iterator that generates the Fibonacci sequence:

In [16]:
class Fibonacci:
    def __init__(self, limit):
        self.limit = limit
        self.current = 0
        self.next = 1

    def __iter__(self):
        return self

    def __next__(self):
        if self.current >= self.limit:
            raise StopIteration
        result = self.current
        self.current, self.next = self.next, self.current + self.next
        return result

fib = Fibonacci(10)
for num in fib:
    print(num)

0
1
1
2
3
5
8


In this example, we define an iterator called `Fibonacci` that generates the Fibonacci sequence up to a given limit. The `__init__()` method initializes the iterator with the limit, the current value, and the next value. The `__iter__()` method returns the iterator object itself, and the `__next__()` method returns the next value in the sequence. The `StopIteration` exception is raised when the iteration is complete.

We create an instance of the `Fibonacci` iterator with a limit of `10`, and then loop over it using a `for` loop. The `for` loop automatically calls the `iter()` function to get the iterator object and the `next()` function to get the next value in the iteration.

Python also provides built-in iterators, such as the `range()` function, that can be used to generate a sequence of numbers. The `range()` function returns an iterator that generates a sequence of numbers from a starting value to an ending value.

### Differnece between Iterator and Generator?

Iterator:

- An iterator is an object that implements the iterator protocol, which consists of the methods `__iter__()` and `__next__()`.
- It allows you to manually iterate over elements using the next() function or by using a for loop.
- Iterators maintain internal state to keep track of the current element.
- Once all the elements have been iterated over, it raises a StopIteration exception to signal the end of the sequence.


Generator:

- A generator is a special type of iterator that is defined using a function with one or more yield statements.
- It generates values lazily, i.e., it produces a value only when requested.
- Generators automatically implement the iterator protocol, so you can iterate over them using a for loop or by using the next() function.
- Unlike iterators, generators automatically manage their internal state, so you don't need to implement `__iter__()` and `__next__()` methods.

<br><br><br>

# Python Generators

A generator in Python is a type of iterator that generates values on-the-fly as they are requested. This means that generators are a more memory-efficient way of generating large sequences of values, as they only generate the values as they are needed, rather than generating them all at once and storing them in memory.

Here's a step-by-step tutorial on how to create and use generators in Python:

### Step 1: Define a generator function using the yield Keyword

A generator function is a special type of function that uses the `yield` keyword to generate values. The `yield` keyword suspends the function's execution and return value to the caller, but it also remembers the function's state so that it can resume execution where it left off the next time it is called.

For example, here's a simple generator function that generates a sequence of numbers:


In [18]:
def number_generator(start, end):
    for num in range(start, end +1):
        yield num

In this example, the `number_generator()` function takes two arguments, `start` and `end`, and generates a sequence of numbers from start to end. The `yield` keyword is used to return each number in the sequence, and to remember the function's state so that it can resume execution where it left off the next time it is called.

### Step 2: Create a generator object using the generator function

To create a generator object, we simply call the generator function. This returns a generator object, which we can use to generate the values on-the-fly.

In [20]:
my_generator = number_generator(1, 5)
my_generator

<generator object number_generator at 0x000002A4FF8274C0>

In this example, we create a generator object called `my_generator` by calling the `number_generator()` function with `start=1` and `end=5`.

### Step 3: Use the generator object to generate the values on-the-fly

We can use the `next()` function to generate the next value in the sequence on-the-fly. Each time we call `next()` on the generator object, the generator function resumes execution where it left off and generates the next value in the sequence.

In [26]:
print(next(my_generator))

StopIteration: 

In this example, we use the `next()` function to generate each value in the sequence on-the-fly, and print it to the console.

### Step 4: Use a for loop to generate all the values in the sequence

We can use a `for` loop to generate all the values in the sequences without having to use the `next()` function explicitly.

In [27]:
my_generator = number_generator(1, 5)
for num in my_generator:
    print(num)

1
2
3
4
5


In this example, we use a `for` loop to generate each value in the sequence on-the-fly, and print it to the console.

### Step 5: Handle the `StopIteration` exception

The `StopIteration` exception is raised when the generator has no more values to generate. We can handle this exception to stop the generation gracefully.

In [28]:
my_generator = number_generator(1, 5)
while True:
    try:
        num = next(my_generator)
        print(num)
    except StopIteration:
        break

1
2
3
4
5


In this example, we use a `while` loop and the `next()` function to generate each value in the sequence on-the-fly, and print it to the console.

The `try` block calls the `next()` function to get the next value in the sequence. If there are no more values, a `StopIteration` exception is raised which is handled in the `except` block by breaking the loop whenever `StopIteration` exception is encountered.

<br><br><br>

# Recursive Function:

Recursive functions are functions that call themselves during their execution. They are useful for solving problems that can be broken down into smaller, similar subproblems. Here's a step-by-step explanation of how recursive functions work in Python.

- **Base Case**: Every recursive function needs a base case, which is the simplest possible case where the function does not call itself. It acts as the termination condition for the recursion. Without a base case, the recursive function would continue calling itself indefinitely, resulting in an infinite loop.

- **Recursive Case**: In addition to the base case, recursive functions also have a recursive case where the function calls itself with a modified version of the problem. This recursive call allows the function to break down a larger problem into smaller subproblems until it reaches the base case.

Now, let's look at an example to illustrate the concept of recursive functions in Python. We'll implement a factorial function using recursion.

In [29]:
def factorial(n):
    if n == 0: #Base case: factorial of 0 is 1
        return 1
    else:
        return n* factorial(n -1) # Recursuve case: n! = n * (n-1)!

print(factorial(5))

120


In this example, the `factorial` function calculates the factorial of a given number `n`. If `n` is equal to `0`, it returns `1`, which is the `base case`. Otherwise, it calls itself with `n-1` and multiplies the result by `n`, which is the `recursive case`.

The function starts with `factorial(5)`. Since `5` is not `0`, it goes to the recursive case and calls `factorial(4)`. Again, `4` is not `0`, so it calls `factorial(3)`. This process continues until `factorial(0)` is reached, which returns 1. The intermediate results are then multiplied together, resulting in `factorial(5) = 5 * 4 * 3 * 2 * 1 = 120`.

Recursive functions can be powerful, but it's important to ensure they have a base case and that the recursive calls converge towards the base case. Otherwise, the function may run indefinitely and cause a stack overflow error.

<br><br><br>